src/Network/GRPC/MQTT/Message/Packet.hs

```diff
@@ -158,55 +164,110 @@ makePacketReader queue = do
+  {-# INLINE displayException #-}
+
+-- | The predicate 'isValidPacketSize' tests if the 'Word32' can be used as the
+-- maximum packet payload size in 'makePacketSender'. This is any 'Word32' value
+-- that is:
+--
+-- * Greater than or equal to 'minPacketSize'.
+--
+-- * Strictly less than 'maxPacketSize'.
+--
+-- @since 1.0.0
+isValidPacketSize :: Word32 -> Bool
+isValidPacketSize x = minPacketSize <= x && x < maxPacketSize
+
+-- | The minimum packet payload size, 16 bytes.
+--
+-- @since 1.0.0
+minPacketSize :: Word32
+minPacketSize = 16
+
+-- | The maximum packet payload size, 256mB.
+--
+-- @since 1.0.0
+maxPacketSize :: Word32
+maxPacketSize = 256 * 2 ^ (20 :: Int)
+
 -- | Construct a packetized message sender given the maximum size for each
 -- packet payload (in bytes), the wire serialization options, and a MQTT publish
 -- function.
 --
 -- @since 0.1.0.0
 makePacketSender ::
   MonadUnliftIO m =>
-  Int ->
+  -- | The maximum packet payload size. This must be a 'Word32' value greater 
+  -- than or equal to 'minPacketSize' and less than 'maxPacketSize'. Otherwise,
+  -- a 'PacketSizeError' exception will be thrown.
+  Word32 ->
   (ByteString -> m ()) ->
   ByteString ->
   m ()
-makePacketSender limit publish message
-  | limit <= 0 || ByteString.length message < limit =
-    -- In the case that the maximum packet payload length @size@ is not at least
-    -- 1 byte or the provided 'ByteString' @bytes@ is empty, then yield one
-    -- terminal packet.
-    let packet :: Packet ByteString
-        packet = Packet message 0 1
-     in publish (wireWrapPacket' packet)
-  | otherwise = do
-    caps <- liftIO getNumCapabilities
-    jobs <- newTVarIO 0
-
-    replicateConcurrently_ caps $ fix \next ->
-      atomically (takeJobId jobs) >>= \case
-        Nothing -> pure ()
-        Just jobid -> do
-          let packet :: Packet ByteString
-              packet = makePacket jobid
-           in publish (wireWrapPacket' packet)
-          next
+makePacketSender limit publish message = do
+  unless (isValidPacketSize limit) do
+    throwIO (PacketSizeError $ toInteger limit)
+
+  if fromIntegral (ByteString.length message) <= payloadSize
+    then do
+      -- In the case that the maximum packet payload length @size@ is not at least
+      -- 1 byte or the provided 'ByteString' @bytes@ is empty, then yield one
+      -- terminal packet.
+      let packet = Packet message 0 1
+      publish (wireWrapPacket' packet)
+    else do
+      caps <- liftIO getNumCapabilities
+      jobs <- newTVarIO 0
+
+      replicateConcurrently_ caps $ fix \next ->
+        atomically (takeJobId jobs) >>= \case
+          Nothing -> pure ()
+          Just jobid -> do
+            let packet = makePacket jobid
+            publish (wireWrapPacket' packet)
+            next
   where
-    njobs :: Int
-    njobs = (ByteString.length message + (limit - 1)) `quot` limit
+    payloadSize :: Word32
+    payloadSize = max (limit - minPacketSize) 1
```

Comment:
Is this correct when `limit == minPacketSize`? It looks like it would end up 1 byte larger than `limit` in that case. Maybe `limit == minPacketSize` shouldn't pass `isValidPacketSize`?

Reply:
If `minPacketSize` and the number bytes introduced from serializing the additional packet fields were exactly equal this would be true. This is why `ByteString.length (writeWrapPacket (Packet "" a b))` is strictly less than `minPacketSize`. I need to add some documentation pointing this out.